File: starter/dsu.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> parent, sz;

int findSet(int v) {
    if (v == parent[v])
        return v;
    return parent[v] = findSet(parent[v]);
}

void unionSet(int a, int b) {
    a = findSet(a);
    b = findSet(b);
    if (a != b) {
        if (sz[a] < sz[b])
            swap(a, b);       
        parent[b] = a;        
        sz[a] += sz[b];
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, q;
    cin >> n >> q;

    parent.resize(n + 1);
    sz.resize(n + 1, 1);

    for (int i = 1; i <= n; i++) {
        parent[i] = i;
    }

    while (q--) {
        char type;
        int a, b;
        cin >> type >> a >> b;

        if (type == '+') {
            unionSet(a, b);
        } else if (type == '?') {
            if (findSet(a) == findSet(b)) cout << "YES\n";
            else cout << "NO\n";
        }
    }

    return 0;
}

File: starter/DSU.java
import java.util.*;

public class DSU {

    static int[] parent, sz;

    static int findSet(int v) {
        // TODO: implement path compression
        return v;
    }

    static void unionSet(int a, int b) {
        // TODO: implement union by size/rank
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int q = sc.nextInt();

        parent = new int[n + 1];
        sz = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            parent[i] = i;
            sz[i] = 1;
        }

        while (q-- > 0) {
            char type = sc.next().charAt(0);
            int a = sc.nextInt();
            int b = sc.nextInt();

            if (type == '+') {
                unionSet(a, b);
            } else { // '?'
                if (findSet(a) == findSet(b)) System.out.println("YES");
                else System.out.println("NO");
            }
        }

        sc.close();
    }
}
